iterative/utils/analytics.go

```diff
@@ -0,0 +1,351 @@
+	"github.com/wessie/appdirs"
+
+	"golang.org/x/crypto/scrypt"
+)
+
+const (
+	Timeout  = 5 * time.Second
+	Endpoint = "https://telemetry.cml.dev/api/v1/s2s/event?ip_policy=strict"
+	Token    = "s2s.jtyjusrpsww4k9b76rrjri.bl62fbzrb7nd9n6vn5bpqt"
+)
+
+var (
+	Version string = "0.0.0"
+	wg      sync.WaitGroup
+)
+
+func getenv(key, defaultValue string) string {
+	value := os.Getenv(key)
+	if len(value) == 0 {
+		return defaultValue
+	}
+	return value
+}
+
+func deterministic(data string) (*uuid.UUID, error) {
+	ns := uuid.NewSHA1(uuid.NameSpaceDNS, []byte("iterative.ai"))
+
+	seed, err := ns.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	dk, err := scrypt.Key([]byte(data), seed, 1<<16, 8, 1, 8)
+	if err != nil {
+		return nil, err
+	}
+
+	id := uuid.NewSHA1(ns, []byte(hex.EncodeToString(dk)))
+	return &id, nil
+}
+
+func SystemInfo() map[string]interface{} {
+	hostStat, _ := host.Info()
+	return map[string]interface{}{
+		"os_name":          hostStat.OS,
+		"platform_version": hostStat.PlatformVersion,
+	}
+}
+
+func TaskDuration(logs string) float64 {
+	regex := regexp.MustCompile(`\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}`)
+	matches := regex.FindAllString(logs, -1)
+	taskDuration := 0.0
+
+	if len(matches) > 1 {
+		layout := "2006-01-02 15:04:05"
+		t1, _ := time.Parse(layout, matches[len(matches)-1])
+		t0, _ := time.Parse(layout, matches[0])
+		taskDuration = t1.Sub(t0).Seconds()
+	}
+
+	return taskDuration
+}
+
+func IsCI() bool {
+	return len(guessCI()) > 0
+}
+
+func guessCI() string {
+	if _, ok := os.LookupEnv("GITHUB_SERVER_URL"); ok {
+		return "github"
+	}
+
+	if _, ok := os.LookupEnv("CI_SERVER_URL"); ok {
+		return "gitlab"
+	}
+
+	if _, ok := os.LookupEnv("BITBUCKET_WORKSPACE"); ok {
+		return "bitbucket"
+	}
+
+	if _, ok := os.LookupEnv("TF_BUILD"); ok {
+		return "azure"
+	}
+
+	if _, ok := os.LookupEnv("CI"); ok {
+		return "unknown"
+	}
+
+	return ""
+}
+
+func TerraformVersion() string {
+	var out bytes.Buffer
+	cmd := exec.Command("terraform", "--version")
+	cmd.Stdout = &out
+
+	err := cmd.Run()
+
+	if err == nil {
+		regex := regexp.MustCompile(`v\d+\.\d+\.\d+`)
+		version := regex.FindString(out.String())
+		return version
+	} else {
+		logrus.Error("Analytics: Failed extracting terraform version")
+	}
+
+	return ""
+}
+
+func GroupId() string {
+	if !IsCI() {
+		return ""
+	}
+
+	rawId := "CI"
+	ci := guessCI()
+
+	if ci == "github" {
+		rawId = fmt.Sprintf("%s/%s",
+			os.Getenv("GITHUB_SERVER_URL"),
+			os.Getenv("GITHUB_REPOSITORY_OWNER"))
+	} else if ci == "gitlab" {
+		rawId = fmt.Sprintf("%s/%s",
+			os.Getenv("CI_SERVER_URL"),
+			os.Getenv("CI_PROJECT_ROOT_NAMESPACE"))
+	} else if ci == "bitbucket" {
+		rawId = os.Getenv("BITBUCKET_WORKSPACE")
+	}
+
+	id, err := deterministic(rawId)
+	if err != nil {
+		return ""
+	}
+
+	return id.String()
+}
+
+func UserId() string {
+	if IsCI() {
+		ci := guessCI()
+		var rawId string
+
+		if ci == "github" {
+			rawId = os.Getenv("GITHUB_ACTOR")
```

Comment:
should be `name login id`, e.g. `Casper da Costa-Luis casperdcl 10780059`

Reply:
Could you please elaborate a bit more? Its not `GITHUB_ACTOR` the right one?

Reply:
`GITHUB_ACTOR` is the username like: `dacbd`, `casperdcl` , or `DavidGOrtega`

Reply:
exactly, we don't want just GITHUB_ACTOR, we want a bit more info too... marginally improves security (harder to reverse the hash)

Reply:
https://github.com/octokit/go-octokit might help :)